src/calculate_bucket.hpp

```diff
@@ -154,69 +159,25 @@ class F1Calculator {
         return std::make_pair(CalculateF(L), L);
     }
 
-    // Returns an evaluation of F1(L), and the metadata (L) that must be stored to evaluate F2,
-    // for 'number_of_evaluations' adjacent inputs.
-    inline std::vector<std::pair<Bits, Bits>> CalculateBuckets(
-        const Bits& start_L,
-        uint64_t number_of_evaluations)
+    // F1(x) values for x in range [first_x, first_x + n) are placed in res[].
+    // n must not be more than 1 << kBatchSizes.
+    void CalculateBuckets(uint64_t first_x, uint64_t n, uint64_t *res)
     {
-        uint16_t num_output_bits = k_;
-        uint16_t block_size_bits = kF1BlockSizeBits;
+        uint64_t start = first_x * k_ / kF1BlockSizeBits;
+        uint64_t end = ((first_x + n + 1) * k_ - 1) / kF1BlockSizeBits;
```

Comment:
there's a pretty well established convention (at least in C++) that "end" refers to the one-past-end position. So to calculate the *number* of entries, you can just do `end - begin`. As far as I can tell, this is the *last* position.

As someone new to the conventions in this code base, it's not all that clear to me what this calculation is doing. The comment above states that `n` is indeed the number of values to be computed, and written to `res`. `end` is only used to compute `n_blks`, which is the number of chacha8 keys to be written to `buf_`, each 64 bytes (`kF1BlockSizeBits` bits).

The intuitive way of computing `end` (I think) would be: `(first_x + n) * k_ / kF1BlockSizeBits`. It looks like what you're doing is a kind of rounding-up, to `k_`, not to `kF1BlockSizeBits`.

in short, I think this computation could use a comment, and either make `end` be one-past-end, or renamed `last`